[Tabs] Add Selection Indicator support.

romoore opened the pull request:
Adds basic support for a selection indicator. Uses simple animation for showing/hiding the indicator as well as repositioning it when the `selectedItem` changes.

![tabbar-selector-portrait](https://user-images.githubusercontent.com/1753199/60371186-84acc400-99c6-11e9-9580-ed4c72cac69d.gif)


Closes #7744

Commit:
[Tabs] Add Selection Indicator support.

Adds basic support for a selection indicator. Uses simple animation for showing/hiding the indicator as well as repositioning it when the `selectedItem` changes.

Closes #7744

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -15,6 +15,9 @@
 #import "MDCTabBarView.h"
 #import "MDCTabBarViewDelegate.h"
 #import "private/MDCTabBarViewItemView.h"
```
clang-format suggested the following change:


```
-#import "private/MDCTabBarViewItemView.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented:
bazel detected changes to the following targets:

```
//components/Tabs:snapshot_tests
//components/Tabs:snapshot_tests_test_bundle
//components/Tabs:unit_tests
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Tabs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Tabs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Tabs:unit_tests_IPHONE_X_IN_11_0
//components/Tabs:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -15,6 +15,9 @@
 #import "MDCTabBarView.h"
 #import "MDCTabBarViewDelegate.h"
 #import "private/MDCTabBarViewItemView.h"
+#import "MDCTabBarViewIndicatorTemplate.h"
```
clang-format suggested the following change:


```
+#import "private/MDCTabBarViewItemView.h"
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -64,10 +71,15 @@ - (instancetype)init {
 
+    _selectionIndicatorTemplate = [MDCTaDefa]
+
     _containerView = [[UIStackView alloc] init];
```
clang-format suggested the following change:


```
-    _containerView = [[UIStackView alloc] init];
+        _containerView = [[UIStackView alloc] init];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -451,6 +463,54 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+  }
+
+
+
```
clang-format suggested the following change:


```
-
-
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -451,6 +463,54 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+  MDCTabBarPrivateIndicatorContext *context =
+  [[MDCTabBarPrivateIndicatorContext alloc] initWithItem:item
+                                                  bounds:selectionIndicatorBounds
+                                            contentFrame:contentFrame];
```
clang-format suggested the following change:


```
-  [[MDCTabBarPrivateIndicatorContext alloc] initWithItem:item
-                                                  bounds:selectionIndicatorBounds
-                                            contentFrame:contentFrame];
+      [[MDCTabBarPrivateIndicatorContext alloc] initWithItem:item
+                                                      bounds:selectionIndicatorBounds
+                                                contentFrame:contentFrame];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -451,6 +463,54 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+  // Ask the template for attributes.
+  id<MDCTabBarIndicatorTemplate> template = _style.selectionIndicatorTemplate;
+  MDCTabBarIndicatorAttributes *indicatorAttributes =
+  [template indicatorAttributesForContext:context];
```
clang-format suggested the following change:


```
-  [template indicatorAttributesForContext:context];
+      [template indicatorAttributesForContext:context];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -451,6 +463,54 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+}
+
+- (CGRect)viewFrameForItemAtIndex:(NSUInteger)index {
+
```
clang-format suggested the following change:


```
-
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.h:
```diff
@@ -33,6 +34,11 @@ __attribute__((objc_subclassing_restricted)) @interface MDCTabBarView : UIScroll
+/**
+ The template for the selection indicator.
+ */
+@property(null_resettable, nonatomic, strong) id<MDCTabBarViewIndicatorTemplate> selectionIndicatorTemplate;
```
clang-format suggested the following change:


```
-@property(null_resettable, nonatomic, strong) id<MDCTabBarViewIndicatorTemplate> selectionIndicatorTemplate;
+@property(null_resettable, nonatomic, strong) id<MDCTabBarViewIndicatorTemplate>
+    selectionIndicatorTemplate;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.h -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented:
## API diff detected the following changes

### Tabs+TabBarView

#### MDCTabBarView

*new* property: `selectionIndicatorTemplate` in `MDCTabBarView`

<!-- identifier: apidiff -->

Commit:
It builds in Xcode!

Commit:
Still going along. Positioning is not right

material-automation commented on components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m:
```diff
@@ -54,7 +54,7 @@ - (void)setUp {
   // Uncomment below to recreate all the goldens (or add the following line to the specific
   // test you wish to recreate the golden for).
-  //  self.recordMode = YES;
+    self.recordMode = YES;
```
clang-format suggested the following change:


```
-    self.recordMode = YES;
+  self.recordMode = YES;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/tests/snapshot/TabBarView/MDCTabBarViewSnapshotTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Saving work. No progress.

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -424,24 +476,55 @@ - (CGFloat)justifiedWidth {
+  }
+
+  if (CGRectGetWidth(self.containerView.bounds) > 0) {
+    NSLog(@"(%@) is located at (%@)", self.items[index].title, NSStringFromCGRect(CGRectStandardize(self.containerView.arrangedSubviews[index].frame)));
```
clang-format suggested the following change:


```
-    NSLog(@"(%@) is located at (%@)", self.items[index].title, NSStringFromCGRect(CGRectStandardize(self.containerView.arrangedSubviews[index].frame)));
+    NSLog(@"(%@) is located at (%@)", self.items[index].title,
+          NSStringFromCGRect(CGRectStandardize(self.containerView.arrangedSubviews[index].frame)));
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -424,24 +476,55 @@ - (CGFloat)justifiedWidth {
+
+  BOOL isRTL =
+      [self mdf_effectiveUserInterfaceLayoutDirection] == UIUserInterfaceLayoutDirectionRightToLeft;
+  CGFloat viewOriginX = isRTL ? CGRectGetMaxX(self.containerView.bounds) : CGRectGetMinX(self.containerView.bounds);
```
clang-format suggested the following change:


```
-  CGFloat viewOriginX = isRTL ? CGRectGetMaxX(self.containerView.bounds) : CGRectGetMinX(self.containerView.bounds);
+  CGFloat viewOriginX =
+      isRTL ? CGRectGetMaxX(self.containerView.bounds) : CGRectGetMinX(self.containerView.bounds);
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -424,24 +476,55 @@ - (CGFloat)justifiedWidth {
+  }
+  CGRect itemFrameInContainerView = CGRectMake(
+      viewOriginX, CGRectGetMinY(self.containerView.bounds), viewSize.width, viewSize.height);
+  CGRect convertedRect =  [self convertRect:itemFrameInContainerView fromView:self.containerView];
```
clang-format suggested the following change:


```
-  CGRect convertedRect =  [self convertRect:itemFrameInContainerView fromView:self.containerView];
+  CGRect convertedRect = [self convertRect:itemFrameInContainerView fromView:self.containerView];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -424,24 +476,55 @@ - (CGFloat)justifiedWidth {
+  if (self.containerView.distribution == UIStackViewDistributionFillEqually) {
+    if (CGRectGetWidth(self.containerView.bounds) < 0) {
+      return CGSizeMake(CGRectGetWidth(self.containerView.bounds) / self.containerView.arrangedSubviews.count,
+                        CGRectGetHeight(self.containerView.bounds));
```
clang-format suggested the following change:


```
-      return CGSizeMake(CGRectGetWidth(self.containerView.bounds) / self.containerView.arrangedSubviews.count,
-                        CGRectGetHeight(self.containerView.bounds));
+      return CGSizeMake(
+          CGRectGetWidth(self.containerView.bounds) / self.containerView.arrangedSubviews.count,
+          CGRectGetHeight(self.containerView.bounds));
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

Commit:
Seems to be working on iPhone 7/iOS 11.2

Commit:
Fix taps on selected item

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -346,18 +392,28 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
+  CGFloat availableWidth = CGRectGetWidth(self.bounds);
+  CGFloat requiredWidth = [self justifiedWidth];
+  UIStackViewDistribution distribution = availableWidth >= requiredWidth ? UIStackViewDistributionFillEqually
+  : UIStackViewDistributionFillProportionally;
```
clang-format suggested the following change:


```
-  UIStackViewDistribution distribution = availableWidth >= requiredWidth ? UIStackViewDistributionFillEqually
-  : UIStackViewDistributionFillProportionally;
+  UIStackViewDistribution distribution = availableWidth >= requiredWidth
+                                             ? UIStackViewDistributionFillEqually
+                                             : UIStackViewDistributionFillProportionally;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

Commit:
Forcing stackview layout

Commit:
Updating snapshot tests

Commit:
Merge branch 'develop' into issue-7744

Commit:
Fix BUILD

material-automation commented:
The following files are not in any BUILD targets:

```
components/Tabs/src/TabBarView/MDCTabBarView.h
components/Tabs/src/TabBarView/MDCTabBarView.m
```


<!-- identifier: bazel-rule-coverage -->

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -352,18 +395,30 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
+    self.containerView.distribution = distribution;
+    [self.containerView setNeedsLayout];
+  }
+  [self.containerView layoutIfNeeded];
```
Is this really needed? I remember it's not good to call layoutIfNeeded in layoutSubviews.

romoore replied:
It's not the best idea, but seems not to cause any errors during testing. Unless we switch to fully manual layout, I'm not sure how else I can get the stack view to reposition its subviews.  I know it's possible to "estimate" the position of those views, but I prefer not to call that code if we can avoid it.

Let me see what breaks when I remove this and I will post back here.

leonmz replied:
According to apple, this method will call LayoutSubviews which could create a calling cycle. The reason that the tests not broken probably because the 

> If no layout updates are pending, this method exits without modifying the layout or calling any layout-related callbacks.

romoore replied:
Calling it from the same view is definitely bad. Calling it from another view is not great.  However, it's not returning while doing nothing - it fixes the layout bugs when I test on the simulator.

I definitely need a hand figuring-out how to get the indicator to appear at the correct location. Forcing a subview to layout isn't the worse option, but it's far from ideal.

Commit:
Merge branch 'develop' into issue-7744

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -374,8 +429,12 @@ - (void)setBounds:(CGRect)bounds {
   if (shouldScroll) {
     [self scrollUntilSelectedItemIsVisibleWithoutAnimation];
+    if (self.selectedItem) {
+      [self updateSelectionIndicatorToIndex:[self.items indexOfObject:self.selectedItem]];
```
Do we also want to call this if shouldScroll is false?

romoore replied:
I will double-check, but I believe that causes problems with scrolling. Scrolling changes the bounds (the `origin`, not the `size`) and I thought animating the selection indicator for each event caused scrolling to break.

romoore replied:
I can't move it out of that check since it breaks layout entirely.

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -457,6 +547,95 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+- (void)updateSelectionIndicatorToIndex:(NSUInteger)index {
+  if (index == NSNotFound || index >= self.items.count) {
+    // Hide selection indicator.
+    self.selectionIndicatorView.bounds = CGRectZero;
```
Maybe use the hidden property to hide the indicator view?

romoore replied:
This is the current logic - I believe it allows animation of the indicator as a "shrinking" effect rather than a fade-out. In either case, it hasn't caused bugs to be filed yet, so I'd prefer to keep it as-is and [the selection indicator animation/layout code can be cleaned-up in the future](https://github.com/material-components/material-components-ios/issues/7748).

mikefan1991 commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -457,6 +547,95 @@ - (CGSize)expectedSizeForView:(UIView *)view {
+
+  // Ask the template for attributes.
+  id<MDCTabBarViewIndicatorTemplate> template = self.selectionIndicatorTemplate;
+  MDCTabBarViewIndicatorAttributes *indicatorAttributes =
```
Why do we need an attributes when we already have the context? It seems increasing the complication.

romoore replied:
I agree that in the future it may be an opportunity to reduce the number of classes used to handle the indicator.  However, it hasn't caused problems so far. I'd rather keep what works (in this case) rather than try to refactor/rewrite it. Given our time limitations, it's probably not the best use of our effort.

Commit:
Merge branch 'develop' into issue-7744

Commit:
Refactoring a bit

romoore commented:
Fails to correctly scroll to the indicator after a rotation.

romoore commented:
@jverkoey I could really use some help on this one if you have time this morning.

Commit:
PR feedback

Commit:
Revert "PR feedback"

This reverts commit 3ebbcf746823bfbed257a2d4462cb02588be7809.

Commit:
[Tabs] Fixes the scrollable content size.

The constraints left the scrollable content size ambiguous in the Justified layout style. Constraining it to the height and width of the content results in no ambiguous layout.

Follow-up to #7753

Commit:
Merge branch 'fix-constraints' into issue-7744

Commit:
Trying constraint-based but to no avail

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -211,6 +244,8 @@ - (void)setSelectedItem:(UITabBarItem *)selectedItem animated:(BOOL)animated {
       [self convertRect:self.containerView.arrangedSubviews[itemIndex].frame
                fromView:self.containerView];
+//  [self didSelectItemAtIndex:itemIndex animateTransition:animated];
+  [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[itemIndex] Animated:animated];
```
clang-format suggested the following change:


```
-//  [self didSelectItemAtIndex:itemIndex animateTransition:animated];
-  [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[itemIndex] Animated:animated];
+  //  [self didSelectItemAtIndex:itemIndex animateTransition:animated];
+  [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[itemIndex]
+                               Animated:animated];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -418,15 +461,28 @@ - (void)observeValueForKeyPath:(NSString *)keyPath
+  [self updateStackviewLayout];
+//  if (self.selectionIndicatorView) {
+//    [self updateSelectionIndicatorToIndex:[self.items indexOfObject:self.selectedItem]];
+//  }
```
clang-format suggested the following change:


```
-//  if (self.selectionIndicatorView) {
-//    [self updateSelectionIndicatorToIndex:[self.items indexOfObject:self.selectedItem]];
-//  }
+  //  if (self.selectionIndicatorView) {
+  //    [self updateSelectionIndicatorToIndex:[self.items indexOfObject:self.selectedItem]];
+  //  }
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -437,24 +493,101 @@ - (void)layoutSubviews {
-    self.initialScrollDone = YES;
-    [self scrollUntilSelectedItemIsVisibleWithoutAnimation];
-  }
+//  [self.containerView layoutIfNeeded];
```
clang-format suggested the following change:


```
-//  [self.containerView layoutIfNeeded];
+  //  [self.containerView layoutIfNeeded];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -437,24 +493,101 @@ - (void)layoutSubviews {
+  CGRect contentFrame = selectionIndicatorBounds;
+  if ([itemView conformsToProtocol:@protocol(MDCTabBarViewIndicatorSupporting)]) {
+    UIView<MDCTabBarViewIndicatorSupporting> *supportingView =
+    (UIView<MDCTabBarViewIndicatorSupporting> *)itemView;
```
clang-format suggested the following change:


```
-    (UIView<MDCTabBarViewIndicatorSupporting> *)itemView;
+        (UIView<MDCTabBarViewIndicatorSupporting> *)itemView;
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -437,24 +493,101 @@ - (void)layoutSubviews {
+  MDCTabBarViewPrivateIndicatorContext *context =
+  [[MDCTabBarViewPrivateIndicatorContext alloc] initWithItem:item
+                                                      bounds:selectionIndicatorBounds
+                                                contentFrame:contentFrame];
```
clang-format suggested the following change:


```
-  [[MDCTabBarViewPrivateIndicatorContext alloc] initWithItem:item
-                                                      bounds:selectionIndicatorBounds
-                                                contentFrame:contentFrame];
+      [[MDCTabBarViewPrivateIndicatorContext alloc] initWithItem:item
+                                                          bounds:selectionIndicatorBounds
+                                                    contentFrame:contentFrame];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -437,24 +493,101 @@ - (void)layoutSubviews {
   if (shouldScroll) {
     [self scrollUntilSelectedItemIsVisibleWithoutAnimation];
+    if (self.selectedItem) {
+[self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[[self.items indexOfObject:self.selectedItem]] Animated:NO];
```
clang-format suggested the following change:


```
-[self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[[self.items indexOfObject:self.selectedItem]] Animated:NO];
+      [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[
+                                                [self.items indexOfObject:self.selectedItem]]
+                                   Animated:NO];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented on components/Tabs/src/TabBarView/MDCTabBarView.m:
```diff
@@ -501,6 +637,9 @@ - (void)updateConstraints {
   }
   self.containerViewConstraintsActive = YES;
+  if (self.selectedItem) {
+    [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[[self.items indexOfObject:self.selectedItem]] Animated:NO];
```
clang-format suggested the following change:


```
-    [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[[self.items indexOfObject:self.selectedItem]] Animated:NO];
+    [self updateSelectionIndicatorForView:self.containerView.arrangedSubviews[
+                                              [self.items indexOfObject:self.selectedItem]]
+                                 Animated:NO];
```

<!-- identifier: clang-format -->
<!-- File: components/Tabs/src/TabBarView/MDCTabBarView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->
